internal/btf/core.go

```diff
@@ -0,0 +1,178 @@
+package btf
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+// Code in this file is derived form libbpf, which is available under a BSD
+// 2-Clause license.
+
+/* The comment below is from bpf_core_types_are_compat in libbpf.c:
+ *
+ * Check local and target types for compatibility. This check is used for
+ * type-based CO-RE relocations and follow slightly different rules than
+ * field-based relocations. This function assumes that root types were already
+ * checked for name match. Beyond that initial root-level name check, names
+ * are completely ignored. Compatibility rules are as follows:
+ *   - any two STRUCTs/UNIONs/FWDs/ENUMs/INTs are considered compatible, but
+ *     kind should match for local and target types (i.e., STRUCT is not
+ *     compatible with UNION);
+ *   - for ENUMs, the size is ignored;
+ *   - for INT, size and signedness are ignored;
+ *   - for ARRAY, dimensionality is ignored, element types are checked for
+ *     compatibility recursively;
+ *   - CONST/VOLATILE/RESTRICT modifiers are ignored;
+ *   - TYPEDEFs/PTRs are compatible if types they pointing to are compatible;
+ *   - FUNC_PROTOs are compatible if they have compatible signature: same
+ *     number of input args and compatible return and argument types.
+ * These rules are not set in stone and probably will be adjusted as we get
+ * more experience with using BPF CO-RE relocations.
+ */
+func coreAreTypesCompatible(localType Type, targetType Type) (bool, error) {
+	var (
+		localTs, targetTs typeDeque
+		l, t              = &localType, &targetType
+		depth             = 0
+	)
+
+	for ; l != nil && t != nil; l, t = localTs.shift(), targetTs.shift() {
+		if depth >= maxTypeDepth {
+			return false, errors.New("types are nested too deep")
+		}
+
+		localType = skipQualifierAndTypedef(*l)
+		targetType = skipQualifierAndTypedef(*t)
+
+		if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
```

Comment:
If we are concerned about performance, we may want to preserve the `BTF_KIND_*` int value and compare those here instead. It seems `reflect.TypeOf` is about 10X the perf of an int comparison: https://play.golang.org/p/h-C4VY2-jdN

```
BenchmarkReflectTypeEqual-8     469669340                2.50 ns/op
BenchmarkIntEqual-8             1000000000               0.249 ns/op
```

Reply:
Interesting, seems like this is because comparing `TypeOf` ends up emitting a call to [runtime.ifaceeq](https://golang.org/src/runtime/alg.go?h=ifaceeq#L295). At this point I'm not terribly worried about the overhead to be honest, especially since we can't take a look what the whole CORE process will look like.